sympy/solvers/ode/systems.py

```diff
@@ -1064,3 +1066,153 @@ def neq_nth_linear_constant_coeff_match(eqs, funcs, t):
+# be capable of dividing the system of ODEs into strongly and
+# weakly connected components. Also, later the argument funcs
+# is needed too.
+def _component_division(eqs, funcs):
+    return [[[eqs, funcs]]]
+
+
+# Returns: List of equations
+def _linear_ode_solver(match):
+    t = match['t']
+    funcs = match['func']
+
+    rhs = match.get('rhs', None)
+    A = -match['func_coeff']
+    B = match.get('commutative_antiderivative', None)
+    const_idx = match['const_idx']
+    type_of_equation = match['type_of_equation']
+
+    sol_vector = linodesolve(A, t, b=rhs, B=B,
+                             type=type_of_equation, const_idx=const_idx)
+
+    sol = [Eq(f, s) for f, s in zip(funcs, sol_vector)]
+
+    return sol
+
+
+# Returns: (List of List of equations, const_idx) or None
+def _ode_component_solver(eqs, funcs, t, const_idx=0):
+    match = neq_nth_linear_constant_coeff_match(eqs, funcs, t)
+    match['const_idx'] = const_idx
+    match['t'] = t
+
+    if match.get('is_linear', False):
+        return [_linear_ode_solver(match)], const_idx + len(eqs)
+    if match.get('is_implicit', False):
+        canon_eqs = match['canon_eqs']
+        sols = []
+        for canon_eq in canon_eqs:
+            sol, const_idx = _ode_component_solver(canon_eq, funcs, t, const_idx=const_idx)
+            if sol is not None:
+                sols += sol
+        return sols, const_idx
+
+    # To add non-linear case here in future
+
+    return None
+
+
+def dsolve_system(eqs, funcs=None, t=None):
+    r"""
+
+    """
+
+    # Note: To add proper preprocessing function
+    # where the function can get the equations from
+    # the expressions.
+
+    eqs = _preprocess_eqs(eqs)
+
+    if any(not isinstance(eq, Eq) for eq in eqs):
+        raise ValueError(filldedent('''
+            List of equations should be passed. The input is not valid.
+        '''))
+
+    # Note: This error checking has to be improved later.
+    if funcs is not None and not isinstance(funcs, list):
+        raise ValueError(filldedent('''
+            Input to the funcs should be a list of functions.
+        '''))
+
+    # Note: This is added to solve a major problem encountered.
+    # Might be best to make a function for this function
+    # extraction later.
+    if funcs is None:
+        funcs = []
+        for eq in eqs:
+            derivs = eq.atoms(Derivative)
+            func = set().union(*[d.atoms(AppliedUndef) for d in derivs])
+            for func_ in func:
+                funcs.append(func_)
+
+    if t is not None and not isinstance(t, Symbol):
+        raise ValueError(filldedent('''
+            The indepedent variable must be of type Symbol
+        '''))
+
+    if t is None:
+        t = list(list(eqs[0].atoms(Derivative))[0].atoms(Symbol))[0]
+
+    sols = []
+    const_idx = 0
+    components = _component_division(eqs, funcs)
+
+    not_solved_systems = []
+
+    for wcc in components:
+        loop_sol = []
```

Comment:
This loop is quite complicated and I don't understand what it is doing. I think it's probably better to refactor this with one or more functions somehow.

Reply:
There should probably be a function for solving a single weakly connected component.

Reply:
@oscarbenjamin Sure thing, I would add such a function with proper comments explaining the blocks of code, as this loop is complicated for me as well. This became complicated since in the last PR, we introduced the fact that our systems solver method can get multiple solutions for an implicit system. Incorporating this with the plan of weakly and strongly connected components made the design lot more complicated.